gcs/COOPERATIVE_LOCKING.md

```diff
@@ -0,0 +1,180 @@
+
+To recover failed directory modification performed with enabled Cooperative
+Locking you should use FSCK tool:
+
+```shell
+hadoop jar /usr/lib/hadoop/lib/gcs-connector.jar \
+  com.google.cloud.hadoop.fs.gcs.CoopLockFsck \
+  --{check|rollBack|rollForward} gs://<bucket_name> [all|<operation-id>]
+```
+
+This command will recover (roll back or roll forward) all failed directory
+modification operations based on operation log.
+
+### Configuration
+
+See
+[Cooperative Locking](CONFIGURATION.md#cooperative-locking-feature-configuration)
+section in [CONFIGURATION.md](CONFIGURATION.md).
+
+### Limitations
+
+Because Cooperative Locking feature is intended to be used by human operators
+when modifying Cloud Storage through `hadoop fs` interface, it has multiple
+limitations that you should take into account before using this feature:
+
+*   It supports isolation of directory modification operations only in the same
+    bucket.
+*   It has a configurable limit (20 by default) of simultaneous directory
+    modification operations per-bucket.
+*   It requires write and list access to the bucket for users that use
+    Cooperative Locking to modify directories in the bucket.
+*   It supports only `hadoop fs` tool and does not support any other Cloud
+    Storage clients (gsutil, API libraries, etc).
+*   To work as intended, if cooperative locking is enabled on one of the clients
+    it also should be enabled on all other clients that modify the same bucket.
+
+### Troubleshooting
+
+To troubleshoot Cooperative Locking feature you can use extensive debug logs
+that can be enabled with `--logglevel` flag:
+
+```shell
+hadoop --loglevel debug fs -Dfs.gs.cooperative.locking.enable=true -{mv|rm} \
+  gs://bucket/path/to/dir1 [gs://gs://bucket/path/to/dir2]
+hadoop --loglevel debug jar /usr/lib/hadoop/lib/gcs-connector.jar \
+  com.google.cloud.hadoop.fs.gcs.CoopLockFsck \
+  --{check|rollBack|rollForward} gs://<bucket_name> [all|<operation-id>]
+```
+
+Also you may want to inspect `all.lock`, `*.lock` and `*.log` files for specific
+directory operations in `_lock/` folder:
+
+*   `all.lock` file
+
+    ```shell
+    gsutil ls -L gs://example-bucket/_lock/all.lock | \
+      grep "lock:" | awk '{print $2}' | base64 --decode | jq
+    ```
+
+    Example output:
+
+    ```json
+    {
+      "formatVersion": 3,
+      "locks": [
+        {
+          "clientId": "your-host.example.com-232396",
+          "operationId": "d265a5f9-bce0-4ccc-b1a7-9e68476f1db8",
+          "operationTime": "2019-08-22T16:26:36.339985Z",
+          "operationType": "RENAME",
+          "lockExpiration": "2019-08-22T16:29:12.396382Z",
+          "resources": [
+            "rename_dst_dir/",
+            "rename_src_dir/"
+          ]
+        }
+      ]
+    }
+    ```
+
+*   Operation `*.lock` file
+
+    ```shell
+    gsutil cat gs://example-bucket/_lock/20190820T163409.446Z_RENAME_2f6e9ca4-0406-4612-82d5-7f07de81aeb0.lock | jq
+    ```
+
+    Example output:
```

Comment:
Briefly introduce the meaning of the sample output.

Reply:
Done.